torch/utils/_foreach_utils.py

```diff
@@ -3,23 +3,23 @@
 
 import torch
 from torch import Tensor
+from torch.autograd.grad_mode import no_grad
 
 
 # This util function splits tensors into groups by device and dtype, which is useful before sending
 # tensors off to a foreach implementation, which requires tensors to be on one device and dtype.
-# Currently, this function is only used in torch.optim.
 # If tensorlistlist contains more than one tensorlist, the following assumptions are made BUT NOT verified:
 #   - tensorlists CAN be None
 #   - all tensors in the first specified list cannot be None
 #   - given an index i, all specified tensorlist[i]s match in dtype and device
-@torch.no_grad()
-def _group_tensors_by_device_and_dtype(tensorlistlist: List[List[Tensor]]) -> Dict[Tuple[str, torch.dtype], List[List[Tensor]]]:
+@no_grad()
+def _group_tensors_by_device_and_dtype(tensorlistlist: List[List[Tensor]]) -> Dict[Tuple[torch.device, torch.dtype], List[List[Tensor]]]:
     assert all([not x or len(x) == len(tensorlistlist[0]) for x in tensorlistlist]), (
            "all specified tensorlists must match in length")
-    per_device_and_dtype_tensors: Dict[Tuple[str, torch.dtype], List[List[Tensor]]] = defaultdict(
+    per_device_and_dtype_tensors: Dict[Tuple[torch.device, torch.dtype], List[List[Tensor]]] = defaultdict(
         lambda: [[] for _ in range(len(tensorlistlist))])
     for i, t in enumerate(tensorlistlist[0]):
-        key = (str(t.device), t.dtype)
```

Comment:
thanks for the cleanup here!

Reply:
np, I hope assuming that the device type will keep being hashable is ok